hicommonwealth/commonwealth
# cosmos_chain_id canonical ID in ChainNodes

## Link to Issue
Closes: #4951 

## Description of Changes
- Adds `cosmos_chain_id` to ChainNodes table
- Adds input for Chain Name to Cosmos Create Community Tab
- Migration for existing chains adds `cosmos_chain_id` based on canonical `chain_name` from [Cosmos Chain Registry](https://github.com/cosmos/chain-registry)
- Validates registration of submitted chain_name against https://cosmoschains.thesilverfox.pro/ API

## "How We Fixed It"


## Test Plan
- CA (click around) tested on local and frack:
  - With an Admin account, go to http://localhost:8080/createCommunity/cosmos
  - Fill out the form:
    - for RPC, input a node endpoint that does not yet exist in the database, like `https://rpc-archway.cosmos-spaces.cloud/` (pulled from https://cosmos.directory/archway/nodes)
    - Fill out Name and Chain Name - confirm validation for chain name (alphanumeric and lowercase)
    - Submit and confirm successful creation (if using a real Cosmos chain_name that we haven't set up yet, like "mayachain")
    - Negative test: Repeat steps above with different rpc (like `https://aura-rpc.tienthuattoan.ventures/`)the _same chain name_ - confirm server error response that Chain Name already exists for a ChainNode
    - Negative test: Repeat steps above with different rpc (like `https://migaloo.rpc.kjnodes.com/`) and a nonsense chain_name like "xyzchain" - confirm server error response that Chain Name is not registered in cosmos.directory registry.

## Deployment Plan

1. 

## Other Considerations

- There are some CosmosSDK ChainNodes in the DB that are not their own chain, or they do not exist in the Chain Registry. I left those unchanged in the migration, so their `cosmos_chain_id` remains null. Ex: Wynd, Polkachu, Injective (Testnet)
- follow-up: https://github.com/hicommonwealth/commonwealth/issues/5063
- I recommend we keep the Cosmos tab admin-only for now. Otherwise, anybody could claim a chain's node as the "official" one.
